site/app/controllers/admin/ReportController.php

```diff
@@ -601,6 +601,41 @@ private function getLateStatusMessage(LateDayInfo $ldi) {
         }
     }
 
+
+
+    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_customization_save", methods: ["POST"])]
+    public function writetocustomization(): JsonResponse {
+        if (!empty($_SERVER['HTTP_X_REQUESTED_WITH']) && strtolower($_SERVER['HTTP_X_REQUESTED_WITH']) == 'xmlhttprequest') {
+            // Build a new model, pull in defaults for the course
+            $customization = new RainbowCustomization($this->core);
+            $customization->buildCustomization();
+
+            if (isset($_POST["json_string"])) {
+                try {
+                    $customization->processForm();
+                    return JsonResponse::getSuccessResponse();
+                }
+                catch (\Exception $e) {
+                    $msg = 'Error processing form';
+                    $this->core->addErrorMessage($msg);
+                    return JsonResponse::getErrorResponse($msg);
+                }
+            }
+            else {
+                $msg = 'No JSON string provided';
+                $this->core->addErrorMessage($msg);
+                return JsonResponse::getErrorResponse($msg);
+            }
+        }
+        else {
+            $msg = 'Not an AJAX request';
+            $this->core->addErrorMessage($msg);
+            return JsonResponse::getErrorResponse($msg);
```

Comment:
There's a lot of duplication here.  Would it be better to move everything but the message text definition to the bottom?

Reply:
good point